### PR TITLE
[FLINK-12676][table][sql-client] Add descriptor, validator, and factory of GenericInMemoryCatalog for table discovery service

### PR DESCRIPTION
**Pull request by bowenli86**
## What is the purpose of the change

This PR adds descriptor, validator, and factory of GenericInMemoryCatalog, and enables SQL CLI to load a generic in memory catalog thru yaml config and table discovery service.

It contains 3 parts:

- commit a754fce8cc91787cf9225a3be166962c4ebafde6 Support 'default-database' config for catalog entries in SQL CLI yaml file
- commit 45742158839b86dda2e0c93a172d9c50b4aae6e2 Add AbstractCatalog to manage the common catalog name and default database name for catalogs
- commit 4c37d371467b54514f91785497af1e3a8b7102c7 Add descriptor, validator, and factory of GenericInMemoryCatalog for table discovery service

## Brief change log

- Added an optional 'default-database' config for catalog entries in SQL CLI yaml file
- Added AbstractCatalog to manage the common catalog name and default database name for catalogs to manage common parts of catalog impls and facilitate comparing catalogs in catalog factory tests
- Developed descriptor, validator, and factory of GenericInMemoryCatalog for table discovery service
- Wrote unit tests for descriptor, validator, and factory of GenericInMemoryCatalog

Descriptor, validator, factory, and unit tests are built by referencing existing classes.

## Verifying this change

This change added tests and can be verified as follows:

- DependencyTest
- ExecutionContextTest.java
- GenericInMemoryCatalogFactoryTest.java 
- GenericInMemoryCatalogDescriptorTest.java

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (docs)

Documents will be added later

### CODE REVIEW

**Comment by bowenli86** [comment]
cc @xuefuz @lirui-apache  @zjuwangg

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/AbstractCatalog.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+package org.apache.flink.table.catalog;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.util.StringUtils;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ * This interface is responsible for reading and writing metadata such as database/table/views/UDFs
```
Comment (xuefuz) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/AbstractCatalog.java]: I think we need to update the comment here.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/AbstractCatalog.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+	public AbstractCatalog(String catalogName, String defaultDatabase) {
+		checkArgument(!StringUtils.isNullOrWhitespaceOnly(catalogName), "catalogName cannot be null or empty");
+		checkArgument(!StringUtils.isNullOrWhitespaceOnly(defaultDatabase), "defaultDatabase cannot be null or empty");
+
+		this.catalogName = catalogName;
+		this.defaultDatabase = defaultDatabase;
+	}
+
+	public String getCatalogName() {
```
Comment (xuefuz) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/AbstractCatalog.java]: While not introduced here, I think we could remove "catalog" from the method and variable name, as these are for Catalog class. The difference is "catalog.getName()" vs "catalog.getCatalogName()", with the former is more natural.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/CatalogDescriptor.java [hunk 3]
```diff
@@ -35,26 +36,49 @@
...
+
+	/**
+	 * Constructs a {@link CatalogDescriptor}.
+	 *
+	 * @param type string that identifies this catalog
+	 * @param propertyVersion property version for backwards compatibility
+	 * @param defaultDatabase default database of the catalog
+	 */
+	public CatalogDescriptor(String type, int propertyVersion, String defaultDatabase) {
 		this.type = type;
```
Comment (xuefuz) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/CatalogDescriptor.java]: check null or empty?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/CatalogDescriptor.java [hunk 4]
```diff
@@ -35,26 +36,49 @@
...
 		return properties.asMap();
 	}
 
+	public String getDefaultDatabase() {
+		return defaultDatabase;
+	}
+
 	/**
 	 * Converts this descriptor into a set of catalog properties.
 	 */
```
Comment (xuefuz) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/CatalogDescriptor.java]: Side q: what's the purpose of this method?

File: flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/gateway/local/ExecutionContextTest.java [hunk 5]
```diff
@@ -73,12 +73,12 @@ public void testExecutionConfig() throws Exception {
...
 	@Test
 	public void testCatalogs() throws Exception {
-		final String catalogName = "catalog1";
+		final String catalogName = "inmemorycatalog";
 		final ExecutionContext<?> context = createCatalogExecutionContext();
 		final TableEnvironment tableEnv = context.createEnvironmentInstance().getTableEnvironment();
 
 		assertEquals(tableEnv.getCurrentCatalog(), catalogName);
-		assertEquals(tableEnv.getCurrentDatabase(), "mydatabase");
+		assertEquals(tableEnv.getCurrentDatabase(), "test-default-database");
```
Comment (xuefuz) [inline flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/gateway/local/ExecutionContextTest.java]: Nit: can we be consistent on how we define the two string variables?

Comment (bowenli86) [reply flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/gateway/local/ExecutionContextTest.java]: reverted

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/AbstractCatalog.java (hunk 2 above)
Comment (bowenli86) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/AbstractCatalog.java]: make sense. renamed it

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/CatalogDescriptor.java (hunk 4 above)
Comment (bowenli86) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/CatalogDescriptor.java]: this pattern is borrowed from existing descriptors. `toProperties()` is for converting common configs for all catalogs, and `toCatalogProperties()` is for converting catalog-specific configs. Impl of `toCatalogProperties()` will always call `super.toProperties()` and then run their own own logic.

**Comment by bowenli86** [comment]
@xuefuz can you take another look?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/AbstractCatalog.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+import org.apache.flink.util.StringUtils;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ * Abstract class for catalogs.
+ */
+@PublicEvolving
+public abstract class AbstractCatalog implements Catalog {
+	private final String catalogName;
```
Comment (xuefuz) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/AbstractCatalog.java]: Nit: catalogName -> name.

**Comment by bowenli86** [comment]
Thanks @xuefuz for your review! I renamed the variable.

Merging

**Pull request closed without merging**